pkg/models/job.go

```diff
@@ -0,0 +1,184 @@
+	// least one active execution.
+	JobStateTypeRunning
+
+	// JobStateTypeCompleted is the state of a job that has successfully completed.
+	// Only valid for batch jobs.
+	JobStateTypeCompleted
+
+	// JobStateTypeFailed is the state of a job that has failed.
+	JobStateTypeFailed
+
+	// JobStateTypeStopped is the state of a job that has been stopped by the user.
+	JobStateTypeStopped
+)
+
+type Job struct {
+	// ID is a unique identifier assigned to this job.
+	// It helps to distinguish jobs with the same name after they have been deleted and re-created.
+	//The ID is generated by the server and should not be set directly by the client.
+	ID string
+
+	// Name is the logical name of the job used to refer to it.
+	// Submitting a job with the same name as an existing job will result in an
+	// update to the existing job.
+	Name string
+
+	// Namespace is the namespace this job is running in.
+	Namespace string
+
+	// Type is the type of job this is, e.g. "daemon" or "batch".
+	Type string
+
+	// Priority defines the scheduling priority of this job.
+	Priority int
+
+	// Count is the number of replicas that should be scheduled.
+	Count int
+
+	// Constraints is a selector which must be true for the compute node to run this job.
+	Constraints []*LabelSelectorRequirement
+
+	// Meta is used to associate arbitrary metadata with this job.
+	Meta map[string]string
+
+	Tasks []*Task
+
+	// State is the current state of the job.
+	State State[JobStateType]
+
+	// Version is a monotonically increasing version number that is incremented
+	// on each job update.
+	Version uint64
+
+	CreateTime int64
+	ModifyTime int64
+}
+
+// NamespacedID returns the namespaced id useful for logging
+func (j *Job) NamespacedID() NamespacedID {
+	return NamespacedID{
+		ID:        j.ID,
+		Namespace: j.Namespace,
+	}
+}
+
+// Normalize is used to canonicalize fields in the Job. This should be
+// called when registering a Job.
+func (j *Job) Normalize() {
+	if j == nil {
+		return
+	}
+
+	// Ensure that an empty and nil map are treated the same to avoid scheduling
+	// problems since we use reflect DeepEquals.
+	if j.Meta == nil {
+		j.Meta = make(map[string]string)
+	}
+
+	// Ensure the job is in a namespace.
+	if j.Namespace == "" {
+		j.Namespace = DefaultNamespace
+	}
+
+	if (j.Type == JobTypeDaemon || j.Type == JobTypeOps) && j.Count == 0 {
+		j.Count = 1
+	}
+
+	for _, task := range j.Tasks {
+		task.Normalize(j)
+	}
+}
+
+// Copy returns a deep copy of the Job. It is expected that callers use recover.
+// This job can panic if the deep copy failed as it uses reflection.
+func (j *Job) Copy() *Job {
+	if j == nil {
+		return nil
+	}
+	nj := new(Job)
+	*nj = *j
+	nj.Constraints = CopySlice[*LabelSelectorRequirement](nj.Constraints)
+
+	if j.Tasks != nil {
+		tasks := make([]*Task, len(nj.Tasks))
+		for i, t := range nj.Tasks {
+			tasks[i] = t.Copy()
+		}
+		nj.Tasks = tasks
+	}
+
+	nj.Meta = maps.Clone(nj.Meta)
+	return nj
+}
```

Comment:
Today we are passing copies of objects all over the place instead of passing pointers, which I understand it is better for safety and data consistency, but not cheep specially for our large Job and Execution types. For example `boltdb` job store deserializes objects from disk and then return a copy of the deserialized object instead of the new object itself. I am hoping that we use more pointers where it makes more sense, and utilize `Copy()` when we want to return a new instance and avoid leaking the original instance. e.g. in-memory data stores can return a copy, and on-disk data stores can return the same deserialized instance